x/rollapp/keeper/invariants.go

```diff
@@ -0,0 +1,204 @@
+				if !found {
+					msg += fmt.Sprintf("rollapp (%s) have no stateInfo at index %d\n", rollapp.RollappId, i)
+					broken = true
+					continue
+				}
+				creationHeight := stateInfo.CreationHeight
+				finalizationTime := creationHeight + k.DisputePeriodInBlocks(ctx)
+
+				val, found := k.GetBlockHeightToFinalizationQueue(ctx, finalizationTime)
+				if !found {
+					msg += fmt.Sprintf("finalizationTime (%d) have no block height\n", finalizationTime)
+					broken = true
+					continue
+				}
+
+				//check that our state index is in the queue
+				found = false
+				for _, idx := range val.FinalizationQueue {
+					if idx.RollappId == rollapp.RollappId && idx.Index == i {
+						found = true
+						break
+					}
+				}
+				if !found {
+					msg += fmt.Sprintf("rollapp (%s) have stateInfo at index %d not in the queue\n", rollapp.RollappId, i)
+					broken = true
+				}
+			}
+		}
+
+		return sdk.FormatInvariant(
+			types.ModuleName, "block-height-to-finalization-queue",
+			msg,
+		), broken
+	}
+}
+
+// RollappCountInvariant checks that the number of rollapps is equal to the number of latestStateInfoIndex
+func RollappCountInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var (
+			broken bool
+			msg    string
+		)
+
+		rollapps := k.GetAllRollapp(ctx)
+		rollappCount := len(rollapps)
+		rollappCountFromIndex := len(k.GetAllLatestStateInfoIndex(ctx))
+
+		if rollappCount == rollappCountFromIndex {
+			return "", false
+		}
+
+		//If the count is not equal, need to check wether it's due to rollapp that didn't publish state yet
+		var noStateRollappCount int
+		for _, rollapp := range rollapps {
+			if !k.IsRollappStarted(ctx, rollapp.RollappId) {
+				noStateRollappCount++
+				continue
+			}
+		}
+
+		broken = rollappCount != (rollappCountFromIndex + noStateRollappCount)
+		if broken {
+			msg = fmt.Sprintf("rollapp count (%d) != latestStateInfoIndex count (%d) + noStateRollapp count (%d)\n", rollappCount, rollappCountFromIndex, noStateRollappCount)
+		}
+
+		return sdk.FormatInvariant(
+			types.ModuleName, "rollapp-count",
+			msg,
+		), broken
+	}
+}
+
+// RollappLatestStateIndexInvariant checks the following invariants per each rollapp that latest state index >= finalized state index
+func RollappLatestStateIndexInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var (
+			broken bool
+			msg    string
+		)
+
+		rollapps := k.GetAllRollapp(ctx)
+		for _, rollapp := range rollapps {
+			if !k.IsRollappStarted(ctx, rollapp.RollappId) {
+				continue
+			}
+
+			latestStateIdx, found := k.GetLatestStateInfoIndex(ctx, rollapp.RollappId)
+			if !found {
+				msg += fmt.Sprintf("rollapp (%s) have no latestStateIdx\n", rollapp.RollappId)
+				broken = true
+			}
+
+			latestFinalizedStateIdx, _ := k.GetLatestFinalizedStateIndex(ctx, rollapp.RollappId)
+			//not found is ok, it means no finalized state yet
+
+			if latestStateIdx.Index < latestFinalizedStateIdx.Index {
```

Comment:
if `latestFinalizedStateIdx` is not found Index should be always 0

Reply:
why? finalized can be `0` but stateIdx can be > 0

Reply:
yea meant that the lastFinalizedStateIndex should always be 0, which I see is what happens when it's not found.